File: crates/kn9t-tui/tests/compactor_ui.lua

function render(state)
  local status = state.status or "unknown"
  local session_id = state.session_id or ""
  local messages_count = state.messages_count or 0
  local tool_calls_count = state.tool_calls_count or 0
  local decisions = state.decisions or {}
  local summary_preview = state.summary_preview or ""
  local triage_done = state.triage_done or false
  local short_id = session_id:sub(1, 8)
  
  -- Status styling
  local status_color = "gray"
  local status_icon = "○"
  local status_text = status
  if status == "reading" then
    status_color = "blue"
    status_icon = "◔"
    status_text = "reading"
  elseif status == "triage" then
    status_color = "yellow"
    status_icon = "↻"
    status_text = "triage"
  elseif status == "triage_retry" then
    status_color = "yellow"
    status_icon = "↻"
    status_text = "triage (retry)"
  elseif status == "summary" then
    status_color = "cyan"
    status_icon = "↻"
    status_text = "summarizing"
  elseif status == "complete" then
    status_color = "green"
    status_icon = "✓"
    status_text = "done"
  elseif status == "error" then
    status_color = "red"
    status_icon = "✕"
    status_text = "failed"
  end
  
  local items = {}
  
  -- Header with status
  local header = status_icon .. " " .. status_text
  if short_id ~= "" then
    header = header .. "  " .. short_id
  end
  table.insert(items, { text = header, fg = status_color })
  
  -- Info line
  if messages_count > 0 then
    local info = tostring(messages_count) .. " msgs"
    if tool_calls_count > 0 then
      info = info .. ", " .. tostring(tool_calls_count) .. " tools"
    end
    table.insert(items, { text = info, fg = "gray" })
  end
  
  -- Tool-call decisions. Kept results are copied verbatim into the summary, so
  -- "keep" is the one the user cares about: show each call by name and command.
  if #decisions > 0 or triage_done then
    table.insert(items, { text = "─── Tool calls (keep = verbatim) ───", fg = "gray" })

    local keep_count = 0
    local summarize_count = 0
    local drop_count = 0
    for _, d in ipairs(decisions) do
      if d.action == "keep" then keep_count = keep_count + 1
      elseif d.action == "summarize" then summarize_count = summarize_count + 1
      elseif d.action == "drop" then drop_count = drop_count + 1
      end
    end

    if #decisions > 0 then
      table.insert(items, {
        text = string.format("keep %d   summarize %d   drop %d",
          keep_count, summarize_count, drop_count),
        fg = "white",
      })

      for _, d in ipairs(decisions) do
        local icon, color, verb = "○", "gray", "?"
        if d.action == "keep" then
          icon, color, verb = "✓", "green", "keep"
        elseif d.action == "summarize" then
          icon, color, verb = "≈", "yellow", "sum"
        elseif d.action == "drop" then
          icon, color, verb = "✕", "red", "drop"
        end
        local name = d.name or d.id:sub(1, 12)
        local line = string.format("%s %-4s %s", icon, verb, name)
        if d.preview and d.preview ~= "" then
          line = line .. "  " .. d.preview
        end
        table.insert(items, { text = line, fg = color })
      end
    end
  end
  
  -- Summary section (shown during/after summary phase)
  if status == "summary" or status == "complete" or summary_preview ~= "" then
    table.insert(items, { text = "─── Summary ───", fg = "gray" })
    if summary_preview ~= "" then
      -- Show full summary, wrapped
      for line in summary_preview:gmatch("[^\\n]+") do
        table.insert(items, { text = line, fg = "cyan" })
      end
    elseif status == "summary" then
      table.insert(items, { text = "generating...", fg = "cyan" })
    end
  end
  
  -- Error message
  if state.error then
    table.insert(items, { text = "⚠ " .. state.error, fg = "red" })
  end
  
  -- The layout owns the frame: it draws the border, the title and the focus
  -- ring around every plugin view. A view returns only its content — wrapping it
  -- in a box here nested two borders and printed two titles.
  return { type = "list", items = items }
end

File: crates/kn9t-tui/tests/compactor_ui.rs
//! The `kn9t-compactor` plugin's Lua viewer, extracted verbatim from its
//! TypeScript source (`scripts/extract_compactor_lua.py`). The plugin lives in the
//! kn9t-plugins repo — https://github.com/96Ems/kn9t-plugins/tree/main/kn9t-compactor.
//!
//! The plugin runs two LLM passes; this only covers what the user sees while it
//! does: which tool calls were kept, summarized or dropped, and by what command.

use kn9t_tui::lua::widgets::Widget;
use kn9t_tui::lua::LuaRuntime;
use kn9t_tui::reducer::{PluginLuaOp, PluginPlacement};
use serde_json::json;

const COMPACTOR_LUA: &str = include_str!("compactor_ui.lua");
const PLUGIN: &str = "kn9t-compactor";

fn runtime_with_compactor(state: serde_json::Value) -> LuaRuntime {
    let rt = LuaRuntime::new().unwrap();
    rt.load_builtin();
    rt.apply_plugin_lua_op(&PluginLuaOp::Register {
        plugin: PLUGIN.into(),
        source: COMPACTOR_LUA.into(),
        placement: PluginPlacement {
            zone: Some("sidebar".into()),
            ..Default::default()
        },
    });
    rt.apply_plugin_lua_op(&PluginLuaOp::SetState {
        plugin: PLUGIN.into(),
        state,
    });
    rt
}

/// Collect every text string in a widget tree.
fn all_text(w: &Widget, out: &mut Vec<String>) {
    match w {
        Widget::Text { .. } => out.extend(w.text()),
        Widget::List { items, .. } => out.extend(
            items
                .iter()
                .map(|spans| spans.iter().map(|s| s.text.as_str()).collect::<String>()),
        ),
        Widget::Box { title, child, .. } => {
            if let Some(t) = title {
                out.push(t.clone());
            }
            if let Some(c) = child {
                all_text(c, out);
            }
        }
        Widget::Float { child: Some(c), .. } => all_text(c, out),
        Widget::Split { children, .. } => {
            for c in children {
                all_text(c, out);
            }
        }
        _ => {}
    }
}

fn texts(rt: &LuaRuntime) -> Vec<String> {
    let w = rt
        .build_plugin_view(PLUGIN)
        .expect("compactor Lua must build");
    let mut out = Vec::new();
    all_text(&w, &mut out);
    out
}

fn state_with_decisions() -> serde_json::Value {
    json!({
        "status": "complete",
        "session_id": "sess-12345678",
        "messages_count": 7,
        "tool_calls_count": 3,
        "triage_done": true,
        "summary_preview": "all done",
        "decisions": [
            {"id": "t1", "action": "keep", "name": "bash", "preview": "git status"},
            {"id": "t2", "action": "drop", "name": "bash", "preview": "ls /tmp"},
            {"id": "t3", "action": "summarize", "name": "read", "preview": "src/main.rs"}
        ]
    })
}

/// The shipped Lua must load.
#[test]
fn shipped_lua_loads_and_renders() {
    let rt = runtime_with_compactor(state_with_decisions());
    assert!(rt.build_plugin_view(PLUGIN).is_ok());
}

/// A kept call must be identifiable: its tool name **and** the command it ran.
#[test]
fn kept_calls_show_the_name_and_the_command() {
    let rt = runtime_with_compactor(state_with_decisions());
    let t = texts(&rt);
    assert!(
        t.iter()
            .any(|s| s.contains("keep") && s.contains("bash") && s.contains("git status")),
        "a kept row must name the tool and its command, got {t:?}"
    );
}

/// The counts line is the at-a-glance answer to "how much is being kept?".
#[test]
fn counts_are_shown_per_action() {
    let rt = runtime_with_compactor(state_with_decisions());
    let t = texts(&rt);
    assert!(
        t.iter()
            .any(|s| s.contains("keep 1") && s.contains("drop 1")),
        "expected a per-action count line, got {t:?}"
    );
}

/// A decision the model left sparse (no name, no preview) must still render.
#[test]
fn sparse_decision_renders_without_error() {
    let rt = runtime_with_compactor(json!({
        "status": "triage",
        "triage_done": true,
        "decisions": [{"id": "abcdef012345", "action": "keep"}]
    }));
    let t = texts(&rt);
    assert!(
        t.iter().any(|s| s.contains("abcdef012345")),
        "a nameless decision falls back to the id, got {t:?}"
    );
}
